### ISSUE TITLE
Minimize use of `fallthrough` in plugins

### ISSUE THREAD

**Issue by miekg** [coredns/coredns#2723]
`fallthrough` is an escape hatch that should not be used lightly. Currently a few plugins use it, some of a good reasons, others not so much:

* *hosts*: per definition of how `/etc/hosts` works, this make sense to allow fallthrough
* *template*: "Continue with the next plugin if the zone matched but no regex matched', looks legit, but can be done with a wildcard match, or an identify function.
* *route53*: "If zone matches and no record can be generated, pass request to the next plugin", looks borderline?
* *etcd*, same text as *route53*, looks *very* borderline
* *kubernetes*, *very* borderline.

The problem with fallthrough is that it makes writing a wrong Corefile easy; this causes extra load and makes a query to be handled by, essentially, the wrong plugin. It also goes against the DNS; as an authoritative plugin (the one that handles the zone) should provide an answer. `fallthrough` encourages plugins to handle 1/2 a zone.

*Hosts* should def. keep it. In *etcd* and *kubernetes* it can be removed. Use in *template* looks ok. I'm not sure about *route53*.

**Comment by chrisohaver** [comment]
> In etcd and kubernetes it can be removed.

In k8s removing `fallthrough` would cause complications in configuration.  `fallthrough` enables us to follow the k8s dns spec by serving PTR records for Endpoint IPs.  At install time, the CIDR for Pods/Endpoints is not known.  Even after instal time, this value isn't trivial to obtain.  Also, in theory, the k8s network addon could expand/change the CIDR range for pods/endpoints at some later point.  Hence the current deployment.

**Comment by chrisohaver** [comment]
Actually correction ... in k8s, `fallthrough` enables us to forward PTR requests upstream if they dont match a k8s object IP.

Removing `fallthrough` _would not result in violating spec_, but it would prevent any PTR requests going upstream.

IIUC, plugins should never be authoritative for reverse zones... unless we make a reverse zone plugin that handles reverse zones only (and create an reverse zones interface for other plugins to implement) ... hmm ...

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Minimize use ..." ]
>Actually correction ... in k8s, `fallthrough` enables us to forward PTR requests upstream if they dont match a k8s object IP.
>
>Removing `fallthrough` _would not result in violating spec_, but it would prevent any PTR requests going upstream.
>
>IIUC, plugins should never be authoritative for reverse zones... unless we make a reverse zone plugin that handles reverse zones only (and create an reverse zones interface for other plugins to implement) ... hmm ...

A reverse zone is not a core DNS concept, it's just a zone. 

Kubernetes is authoritative for the reverse of it's cluster, nothing more.

~~~
cluster.local 10.244/16 {
    config
}
. {
    forward
    cache
    etc
}
~~~

is a more to the point configuratiom.

I get that configuration is easy by allowing it to be more generic. But this is 
a one time thing and you're payingback for it with every query that is handled 
by plugin that should not handle it. The configuration itself can also be 
automated.

**Comment by chrisohaver** [comment]
The interim solution in K8s would to ask the user to specify their pod CIDR range during installation time, and if they do not know it, leave it blank (resulting in spec non-compliance).

It's only PTRs so I guess it's not a huge deal for most people.

**Comment by dilyevsky** [comment]
For route53 it is not uncommon to have two zones with the same name where one is private (https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/hosted-zones-private.html) and the other one is public (perhaps not even served by route53 at all) so fallthrough is definitely advantageous for cases like this as regular `upstream` wouldn't cover this.

**Comment by johnbelamaric** [comment]
Configuration of the reverse zones in Kubernetes is not a one-time thing. It can change over time - for example, some clusters will allocate a CIDR for pods on a node when the node is added. Nodes can be added or removed at any time - even in an automated way with cluster API or other cluster autoscalers. On top of that, this information is not available programatically. Without fallthrough, PTR requests are broken whenever you autoscale, unless you insert a manual step of reconfiguring the DNS service or provide some bespoke orchestration logic to reconfigure the DNS server. I don't think that is tenable.

**Comment by ramnes** [comment]
In our situation, we have an internal domain to which we want to give control to developers with ExternalDNS, on several Kubernetes cluster. The main Kubernetes cluster is on-premise, others are in AWS, and two different teams own the two different setups. We would like to have all ExternalDNS instances in all Kubernetes clusters to speak to our company's main DNS server, but it's an old and non-standard technology that isn't supported by ExternalDNS.

Our plan is indeed to migrate to something newer and standard, but until then we found a temporary solution in registering the subdomains of the main Kubernetes cluster in etcd and serve them through CoreDNS, while having the few ones of AWS clusters in route53, and falling through between the two setups.

TL;DR: `fallthrough` is indeed risky and hackish, but still a good feature of CoreDNS as it allows to quickly answer boring edge cases, IMHO.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Minimize use ..." ]
>In our situation, we have an internal domain to which we want to give control to developers with ExternalDNS, on several Kubernetes cluster. The main Kubernetes cluster is on-premise, others are in AWS, and two different teams own the two different setups. We would like to have all ExternalDNS instances in all Kubernetes clusters to speak to our company's main DNS server, but it's an old and non-standard technology that isn't supported by ExternalDNS.
>
>Our plan is indeed to migrate to something newer and standard, but until then we found a temporary solution in registering the subdomains of the main Kubernetes cluster in etcd and serve them through CoreDNS, while having the few ones of AWS clusters in route53, and falling through between the two setups.
>
>TL;DR: `fallthrough` is indeed risky and hackish, but still a good feature of CoreDNS as it allows to quickly answer boring edge cases, IMHO.

thanks. I think your tl;dr captures my concerns well. Probably worth putting 
that in the docs of the fallthrough package.
